### ISSUE TITLE
Image stretch when Window Resize is enabled

### ISSUE THREAD

**Issue by drmar** [godotengine/godot#37288]
**Godot version:**
3.2.1

**OS/device including version:**
macOS Catalina 10.15.4 / MacBookPro11,1

**Issue description:**
If Window Resize is enabled, the image gets stretched vertically.

![Screenshot 2020-03-25 at 11 54 35](https://user-images.githubusercontent.com/13488692/77524196-9953b500-6e8f-11ea-82f0-9780ef984e18.png)
![Screenshot 2020-03-25 at 11 54 49](https://user-images.githubusercontent.com/13488692/77524210-9f499600-6e8f-11ea-8124-ff5dd0869116.png)

**Note:**
I set the Environment color to white, but when the image is stretched the borders are black, so it is not showing the environment color. Something else must be happening.

![Screenshot 2020-03-25 at 11 55 02](https://user-images.githubusercontent.com/13488692/77524323-d455e880-6e8f-11ea-80a2-7806d40e614b.png)

**Steps to reproduce:**
Run the attached Godot project with and without Window Resize enabled.

**Minimal reproduction project:**
[Godot_Window_Bug.zip](https://github.com/godotengine/godot/files/4380118/Godot_Window_Bug.zip)

**Comment by Ev1lbl0w** [comment]
I can't reproduce this issue, probably because I only have a *1366x768* screen :stuck_out_tongue: 

Regarding the black bars, I don't think it's an issue. The clear color refers to the viewport's clear color; what's outside of the viewport becomes black. You can see this by creating a blank scene, and seeing both the black and white stripes when you resize:
![image](https://user-images.githubusercontent.com/6501975/77534637-ee43fb00-6e90-11ea-9d93-b192172eb99c.png)

I'm not sure if you can modify the color of the area outside the viewport.

**Comment by drmar** [comment]
@Ev1lbl0w Nevermind the color of the Environment. I only mentioned it as an observation, in case it would help with debugging. That is not the point.

The problem is that the image gets stretched horizontally when Window Resize is enabled.

Also, I think it may be related to a broader set of bugs related to rendering and windowing. I reported several bugs related to those issues.

**Comment by Ev1lbl0w** [comment]
So, just to be clear; the two screenshots you posted happen when **Window Resize** is disabled, and enabled, respectively?

**Comment by drmar** [comment]
> So, just to be clear; the two screenshots you posted happen when Window Resize is disabled, and enabled, respectively?

Yes. The only difference is Window Resize enabled and disabled, respectively. No other changes were made to the project.

**Comment by Ev1lbl0w** [comment]
Try printing the resolution in both cases, and comparing them:

```gdscript
func _ready():
    print(OS.window_size)
```

**Comment by drmar** [comment]
@Ev1lbl0w You are on to something!

**Resolution output**
With Resizable DISABLED: (1920, 1080)
With Resizable ENABLED: (1920, 1058)

**Comment by Calinou** [comment]
@drmar This might be due to window borders and the task bar affecting the final resolution when the window is set to be resizable.

**Comment by drmar** [comment]
@Calinou 

> This might be due to window borders and the task bar affecting the final resolution when the window is set to be resizable.

I don't think this is the case. The same problem occurs both in fullscreen and in window screen.

**Comment by Ev1lbl0w** [comment]
> I don't think this is the case. The same problem occurs both in fullscreen and in window mode.

Do you mean that setting it to fullscreen and resizable also causes this issue?

**Comment by drmar** [comment]
@Ev1lbl0w 
> Do you mean that setting it to fullscreen and resizable also causes this issue?

I mean that it is irrelevant if you run the game with the Fullscreen option enabled or disabled.
The same exact horizontal stretching problem occurs, which leads me to conclude that the boarders of the screen and the Status Bar/Menu have nothing to do with it.

I would encourage you to simply download the project yourselves and run it.

**Comment by Ev1lbl0w** [comment]
> I mean that it is irrelevant if you run the game with the Fullscreen option enabled or disabled.
> The same exact horizontal stretching problem occurs, which leads me to conclude that the boarders of the screen and the Status Bar/Menu have nothing to do with it.
> 
> I would encourage you to simply download the project yourselves and run it.

Yes, I have tried to run the project, but like I said, I can't reproduce this issue. I have tried different combinations of sizes, fullscreen and window resize, but haven't had any luck finding the problem.

So, in that case, does the resolution **also** change in fullscreen mode? Because that should be impossible AFAIK.

Please post the output from this code:

```gdscript
func _ready():
	print("Screen:   ", OS.get_screen_size())
	print("Window:   ", OS.window_size)
	print("Viewport: ", get_viewport().size)
```

In all situations possible *(with/without fullscreen, with/without resize)*

**Comment by drmar** [comment]
Alright. I will even make a video for you. Please allow me a few minutes.

**Comment by drmar** [comment]
@Ev1lbl0w Here is the download link for the screen recording, that includes with/without fullscreen, with/without resize:
https://drive.google.com/open?id=1rfRweWs5Yo0MgLJRYsfnOl7Uj8REgdKI

Let me know if the problem is not clear to you.

**Comment by drmar** [comment]
In text form, here are the outputs for the various combinations:

**WITHOUT Fullscreen / WITHOUT Resizable:**
Screen:   (1920, 1080)
Window:   (1920, 1080)
Viewport: (1920, 1080)

**WITH Fullscreen / WITH Resizable:**
Screen:   (1920, 1080)
Window:   (1920, 1058)
Viewport: (1880, 1058)

**WITH Fullscreen / WITHOUT Resizable:**
Screen:   (1920, 1080)
Window:   (1920, 1080)
Viewport: (1920, 1080)

**WITHOUT Fullscreen / WITH Resizable:**
Screen:   (1920, 1080)
Window:   (1920, 1058)
Viewport: (1880, 1058)

**Comment by Ev1lbl0w** [comment]
Thank your for the outputs and the video, it was very useful :wink:

I believe that, in window mode, this isn't a problem. The window that is (1920, 1080) (no fullscreen, no resizable) is probably bigger than the screen, and if you drag it up, you might find the rest of it that was hidden. *(but if it isn't, then it's probably an issue too)*

The fullscreen, however, looks like an issue. I believe it shouldn't have any other resolution other than the screen size, yet it has a lower resolution.

I'll see if I can dig anything. I might get access to a Mac machine to test this, but can't make any promises :stuck_out_tongue:

**Comment by drmar** [comment]
@Ev1lbl0w No problem! I am doing my part to help Godot get better and better.

Regarding window mode, I beg to differ, honestly. The same exact problem happens in both full screen and in window mode. I honestly don't believe the having Fullscreen option enabled or disabled changes anything related to this specific problem.

However, what I do believe creates this particular problem is the TextureRect.
If you recreate the same project with Sprite instead of TextureRect, the problem doesn't occur anymore. Please note that this is just speculation, I am not an expert by any stretch.

**Comment by Ev1lbl0w** [comment]
Hey, I got some news. So I was able to try this on a Mac from a friend of mine. The screen size was (2048, 1152), so not the 1080p I was expecting.

I tried running your project in all the configurations and tried to replicate the problem, but couldn't. When it was in fullscreen mode, having resized on or off didn't matter: the image wasn't being stretched.

So, I don't know what's going on :confused:

**Comment by nisovin** [comment]
Are you sure the image is getting stretched? It seems more like it's getting scaled. When you have resizable off, the window doesn't fit in your screen, so the bottom part of the image disappears off the bottom. With resizable on, the window is allowed to be resized to fit the screen, and due to your stretch mode settings, the left and right black bars appear to maintain the aspect ratio.

**Comment by drmar** [comment]
@nisovin The image should fit perfectly on screen, since it is precisely 1920x1080, which is the same resolution as the monitor I'm using.

**Comment by pwnorbitals** [comment]
@drmar is this still happening ?

**Comment by drmar** [comment]
> @drmar is this still happening ?

In Godot 3.3.2.stable the problem is still happening.

**Comment by starryalley** [comment]
I see similar issue in Godot 3.3.4 stable. I always see window and viewport report a bit less in y than the actual screen size when in fullscreen. 

Also there is a weird issue about rendering 2d image on a 1792x960 resolution screen on Mac. Project's width/height is also 1792x960. If "fullscreen" and "resizable" is both checked, there are some obvious artifacts on 2d images (of resolution also 1792x960). If "fullscreen" is checked, but "resizable" is unchecked, the image is rendered perfectly. See the screenshots attached which is part of the image (black background, some white text). 

I didn't check in non-fullscreen case though. 

![image](https://user-images.githubusercontent.com/1099404/201000070-4df3d51b-75e9-4bd4-9bcd-4a511e858675.png)
![image](https://user-images.githubusercontent.com/1099404/201000116-02f6bd46-af14-4da3-98ee-47cf608d976a.png)

**Comment by Calinou** [comment]
@starryalley Are you using the `2d` or `viewport` stretch mode? If using the `2d` stretch mode, this sounds like the font oversampling bug that has already been reported.

Also, [Godot 3.3.x is not supported for bug fixes anymore](https://docs.godotengine.org/en/stable/about/release_policy.html#release-support-timeline), so please upgrade to [3.5.1](https://godotengine.org/article/maintenance-release-godot-3-5-1) and check if the issue still occurs there.

**Comment by starryalley** [comment]
No, stretch mode is disabled. The project is a 2d only project with all 2d images exactly matching the screen resolution and it runs fullscreen. Sorry I might not be clear enough. The screenshots above were from a 1792x960 image rendered by godot.

I'll find some time to try 3.5.1 and report back since this only happens on a work-related machine which I'm not always accessible. My setup with a standard 2k (2560x1440) monitor doesn't have this issue though.
